src/main/java/org/apache/sysml/parser/DMLTranslator.java

```diff
@@ -267,10 +266,20 @@ public void rewriteHopsDAG(DMLProgram dmlp)
 		//propagate size information from main into functions (but conservatively)
 		if( OptimizerUtils.ALLOW_INTER_PROCEDURAL_ANALYSIS ) {
 			InterProceduralAnalysis ipa = new InterProceduralAnalysis();
-		    ipa.analyzeProgram(dmlp);
-		   	resetHopsDAGVisitStatus(dmlp);
+			ipa.analyzeProgram(dmlp);
+			resetHopsDAGVisitStatus(dmlp);
+			if (OptimizerUtils.ALLOW_IPA_SECOND_CHANCE) {
+				// SECOND CHANCE:
+				// Rerun static rewrites + IPA to allow for further improvements, such as making use
+				// of constant folding (static rewrite) after scalar -> literal replacement (IPA),
+				// and then further scalar -> literal replacement (IPA).
+				rewriter.rewriteProgramHopDAGs(dmlp);
+				resetHopsDAGVisitStatus(dmlp);
+				ipa.analyzeProgram(dmlp);
+				resetHopsDAGVisitStatus(dmlp);
+			}
```

Comment:
Well I think it might be cleaner to leave it up at this level because other wise I'd have to pass in the `rewriter` into `InterProceduralAnalysis.analyzeProgram`, or create another one, both of which would just get messy.  I removed the TODO message.  Does that work?